### PR TITLE
Deprecate `pvsystem.singlediode` parameter `ivcurve_pnts`

### PR DESCRIPTION
**Pull request by reepoi**
- [x] Partially addresses #1626 (last three checkboxes), [comment one](https://github.com/pvlib/pvlib-python/issues/1626#issuecomment-1363345991), and [comment two](https://github.com/pvlib/pvlib-python/issues/1626#issuecomment-1363385583).
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

The `ivcurve_pnts` parameter of `pvsystem.singlediode` is deprecated in favor of using `pvsystem.i_from_v` and `v_from_i` to get additional points on an IV curve. The test cases and `docs/examples/iv-modeling/plot_singlediode.py` are updated. Also, a bug in `tools._golden_sect_DataFrame` where scalars were converted into 0d-arrays is fixed.

`pvsystem.singlediode` returns a `dict` or a `pd.DataFrame`. A `dict` is returned only when all inputs are scalars or `ivcurve_pnts > 0`.

### CODE REVIEW

File: pvlib/pvsystem.py [hunk 1]
```diff
@@ -2844,22 +2841,25 @@ def singlediode(photocurrent, saturation_current, resistance_series,
        photovoltaic cell interconnection circuits" JW Bishop, Solar Cell (1988)
        https://doi.org/10.1016/0379-6787(88)90059-2
     """
+    if ivcurve_pnts:
```
Comment (cwhanse) [inline pvlib/pvsystem.py:2844]: Probably should add a test in `test_pvsystem` to check that this warning is emitted, if nothing else, to ensure coverage of these lines.

**Review by cwhanse — APPROVED** [review]

File: docs/examples/iv-modeling/plot_singlediode.py [hunk 2]
```diff
@@ -32,6 +32,7 @@
 # :py:meth:`pvlib.pvsystem.singlediode` is then used to generate the IV curves.
```
Comment (cwhanse) [inline docs/examples/iv-modeling/plot_singlediode.py:32]:
```suggestion
# :py:meth:`pvlib.pvsystem.singlediode` and :py:meth:`pvlib.pvsystem.i_from_v` are used to generate the IV curves.
```

**Review by cwhanse — APPROVED** [review]
CI failure is codecov, ignorable.

**Review by kandersolar — COMMENTED** [review]
Sorry for the late review, I missed that this was tagged 0.10.0!  Looks good to me overall; the only thing I see is that these deprecation warnings from the tests need to get cleaned up: https://github.com/pvlib/pvlib-python/actions/runs/5201899928/jobs/9382722429?pr=1743#step:9:69

**Comment by reepoi** [comment]
Ok, I wrapped all the tests using `pvsystem.singlediode`'s `ivcurve_pnts` parameter with `with pytest.warns` to hide the warning showing in the pytest output.

File: pvlib/tests/ivtools/test_sde.py [hunk 3]
```diff
@@ -11,12 +12,13 @@ def get_test_iv_params():
...
 def test_fit_sandia_simple(get_test_iv_params, get_bad_iv_curves):
     test_params = get_test_iv_params
-    testcurve = pvsystem.singlediode(photocurrent=test_params['IL'],
-                                     saturation_current=test_params['I0'],
-                                     resistance_shunt=test_params['Rsh'],
-                                     resistance_series=test_params['Rs'],
-                                     nNsVth=test_params['nNsVth'],
-                                     ivcurve_pnts=300)
+    with pytest.warns(pvlibDeprecationWarning, match='ivcurve_pnts'):
+        testcurve = pvsystem.singlediode(photocurrent=test_params['IL'],
```
[the remark this answers is not in the document]
Comment (reepoi) [reply pvlib/tests/ivtools/test_sde.py]: That makes sense. I've updated these two test cases.

File: pvlib/tests/test_pvsystem.py [hunk 4]
```diff
@@ -1426,6 +1429,13 @@ def test_singlediode_series_ivcurve(cec_module_params):
         assert_allclose(v, expected[k], atol=1e-6)
 
 
+@pytest.mark.parametrize('method', ['lambertw', 'brentq', 'newton'])
```
Comment (kandersolar) [inline pvlib/tests/test_pvsystem.py:1432]:
```suggestion
@fail_on_pvlib_version('0.11')
@pytest.mark.parametrize('method', ['lambertw', 'brentq', 'newton'])
```

To remind us to remove this when the deprecation period is over

**Review by kandersolar — APPROVED** [review]
LGTM.  Thanks @reepoi!

**Pull request merged**